ItemLevelDownloader/Engine/fetchEngine.ts

```diff
@@ -0,0 +1,119 @@
+                            if (!fs.existsSync(folder)) {
+                                fs.mkdir(folder);
+                            }
+                            createdFolders[folder] = true;
+                        }
+
+                        await this.downloadArtifactItem(artifactProvider, item, outputFilename, fetchEngineOptions.downloadPattern, fetchEngineOptions.retryLimit);
+                    }
+
+                    console.log("Exiting worker nothing more to process");
+
+                    resolve();
+                }
+                catch (err) {
+                    reject(err);
+                }
+            }));
+        }
+
+        await Promise.all(downloaders);
+    }
+
+    downloadArtifactItem(artifactProvider: models.IArtifactProvider,
+        item: models.ArtifactItem,
+        outputFilename: string,
+        downloadPattern: string,
+        retryLimit: number,
+        retryCount?: number): Promise<{}> {
+        return new Promise(async (downloadResolve, downloadReject) => {
+            try {
+                retryCount = retryCount ? retryCount : 0;
+                if (item.itemType === models.ItemType.File) {
+                    if (minimatch(item.path, downloadPattern)) {
+                        console.log("Downloading '%s' to '%s' (file %d of %d)", item.path, outputFilename);
+                        const contentStream = await artifactProvider.getArtifactItem(item);
+                        const outputStream = fs.createWriteStream(outputFilename);
+
+                        contentStream.pipe(outputStream);
+                        contentStream.on("end",
+                            () => {
+                                console.log(`Downloaded '${item.path}' to '${outputFilename}'`);
+                                downloadResolve();
+                            });
+                    }
+                    else {
+                        console.log("Skipping download of file " + item.path);
+                        downloadResolve();
+                    }
+                }
+                else {
+                    var items = await artifactProvider.getArtifactItems(item);
+                    items = items.map((value, index) => {
+                        if(!value.path.startsWith(item.path)){
+                            value.path = path.join(item.path, value.path);
+                        }
+
+                        return value;
+                    });
+
+                    this.artifactStore.addItems(items);
+
+                    console.log("Enqueued " + items.length + " to download queue for processing.");
+                    downloadResolve();
+                }
+            } catch (err) {
+                console.log("Error downloading file %s: %s", item.path, err);
+                if (retryCount === retryLimit - 1) {
+                    downloadReject(err);
+                } else {
+                    process.nextTick(() => this
```

Comment:
for some reason I needed to add the download to the end of current event-loop, I will add supporting retryIntervalInSeconds in backlog and go over it later.